explainers/subresource-loading.md

```diff
@@ -0,0 +1,101 @@
+  with tree shaking could invalidate everything).
+
+- With JS bundlers, execution needs to wait for the full bytes to come. Ideally
+  loading multiple subresources should be able to utilize full streaming and
+  parallelization, while it's not possible if all resources are bundled as one
+  javascript. (For JS modules execution still needs to be waited for the entire
+  tree due to the current
+  [deterministic execution model](https://docs.google.com/document/d/1MJK0zigKbH4WFCKcHsWwFAzpU_DZppEAOpYJlIW7M7E/edit#heading=h.5652gd5ks5id))
+
+- Related issues: [#411](https://github.com/WICG/webpackage/issues/411),
+  [#526](https://github.com/WICG/webpackage/issues/526)
+
+## Proposal
+
+We propose a new `link rel=”webbundle”` notion [**1**] that lets the browser
+fetch bundled resources and attach them to the page. Such as:
+
+```html
+<link
+  rel="webbundle"
+  href="https://example.com/dir/subresources.wbn"
+  scope="https://example.com/scope/"
+/>
+```
+
+If the main document has such a `link` element, a browser must do:
+
+- A browser must fetch the specified Web Bundle, `subresources.wbn`.
+
+- All the subresources whose URLs are under the scope
+  `https://example.com/scope/` must [**2**] be loaded from the specified Web
+  Bundle.
+
+[**1**] The syntax is tentative. The API surface is not so important for us yet.
+Any alternative ideas are welcome. This is basically about mapping a set of
+resources to a particular bundled resource. Something like
+[Fetch maps proposal](https://discourse.wicg.io/t/proposal-fetch-maps/4259)
+might be able to extend to this case.
+
+[**2**] The exact behavior is TBD. There are several possible behaviors when the
+bundle does not contain the subresource; e.g. just fail, fallback to the
+network, or allow users to specify its behavior with a `link` element's
+attribute.
+
+## Example
+
+### The bundle
+
+Suppose that the bundle, `subresources.wbn`, includes the following resources:
+
+```
+- https://example.com/scope/a.js (which depends on ./b.js)
+- https://example.com/scope/b.js
+- https://example.com/scope/c.png
+- … (omitted)
+```
+
+### The main document
+
+```html
+<link rel="webbundle"
+    href="https://example.com/dir/subresources.wbn"
+    scope="https://example.com/scope/"
+/>
+
+<script type=”module” src=”https://example.com/scope/a.js”></script>
+<img src=https://example.com/scope/c.png>
+```
+
+Then, a browser must fetch the bundle, `subresources.wbn`, and load
+subresources, `a.js`, `b.js`, and `c.png`, from the bundle.
+
+## Subsequent loading and Caching
+
+On the subsequent navigations for the same page/origin, it’s desirable that the
+resources that are already fetched in the bundle shouldn't be refetched. We're
+thinking about two possibilities here:
+
+1. Embed the details of version / cached resources in the URL
+2. Use a cache digest or something (and also embed it in the request)
+
+Then the server should be able to dynamically subset the Bundle and serve it.
```

Comment:
A patch format also seems plausible. If the server looks at the If-None-Match request header to learn what version the client already has cached, it can send down a different response type to patch the bundle instead of sending down the whole bundle. I don't know of any existing systems that do that, so we should talk to more people if that direction sounds good. Roberto Peon has argued for patches.

Reply:
Thanks! Yes, we are currently discussing this topic actively in other places. I think we'll cover this topic in another explainer!